### PR TITLE
[WIP] Notifications

### PR DESCRIPTION
**Pull request by GrahamCampbell**
DO NOT MERGE THIS YET. IT'S TOTALLY UNREADY!

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
@Crash-- You've misused migrations here. That belongs in a seeder.

File: composer.json [hunk 1]
```diff
@@ -8,6 +8,12 @@
     ],
     "license": "MIT",
     "type": "project",
+    "repositories": [
+        {
+            "type": "git",
+            "url": "https://github.com/Sjimi/slack"
```
Comment (GrahamCampbell) [inline composer.json]: we don't even need this fork?

File: src/Http/Controllers/NotificationController.php [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        ];
+
+        View::share('subTitle', $this->subTitle);
+        View::share('subMenu', $this->subMenu);
+    }
+
+    /**
+     * Show Slack's configuration.
+     *
+     * @return View
```
Comment (GrahamCampbell) [inline src/Http/Controllers/NotificationController.php]: incorrecrt return type, we are not returning illuminate\support\facades\view...

File: src/Http/Controllers/NotificationController.php [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            'partial'    => 'slack',
+            'properties' => $slackProperties,
+            'active'     => $slackConfiguration->active,
+            ]);
+    }
+
+    /**
+     * Show Twilio's configuration.
+     *
+     * @return View
```
Comment (GrahamCampbell) [inline src/Http/Controllers/NotificationController.php]: same

File: src/Http/Controllers/NotificationController.php [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            'partial'    => 'twilio',
+            'properties' => $twilioProperties,
+            'active'     => $twilioConfiguration->active,
+        ]);
+    }
+
+    /**
+     * Edit Slack configuration.
+     *
+     * @return Redirect
```
Comment (GrahamCampbell) [inline src/Http/Controllers/NotificationController.php]: same

File: src/Http/Controllers/NotificationController.php [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+        $slackConfiguration->save();
+
+        return Redirect::back();
+    }
+
+    /**
+     * Edit Twilio configuration.
+     *
+     * @return Redirect
```
Comment (GrahamCampbell) [inline src/Http/Controllers/NotificationController.php]: same

File: src/Models/Observers/IncidentObserver.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+use CachetHQ\Cachet\Services\Notifications\NotifierAbstract;
+
+class IncidentObserver
+{
+    /**
+     * When an Incident is saved (created or updated) this event is triggered.
+     *
+     * @param \Illuminate\Database\Eloquent\Model $model
+     *
+     * return void
```
Comment (GrahamCampbell) [inline src/Models/Observers/IncidentObserver.php]: incorrect

File: src/Models/Observers/IncidentObserver.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        $services = Service::where('active', 1)->get();
+
+        foreach ($services as $service) {
+            $properties = $service->properties;
+
+            $notifier = new NotifierAbstract();
+            $classname = 'CachetHQ\\Cachet\\Services\\Notifications\\'.$properties->notifierName;
+            $notifier->setNotifier(new $classname())
+                     ->setParamsToNotifier($properties)
```
Comment (GrahamCampbell) [inline src/Models/Observers/IncidentObserver.php]: alignment all wrong

**Comment by Crash--** [comment]
@GrahamCampbell OK but I absolutely need these data in the DB. Without, we'll have a problem, this is why I put it into a migration because migrations are mandatory and seeds are optional, right?

File: src/Notifications/TwilioNotifier.php [hunk 8]
```diff
@@ -0,0 +1,84 @@
+<?php
+
+namespace CachetHQ\Cachet\Notifications;
+
+use Illuminate\Database\Eloquent\Model;
+use Services_Twilio as Twilio;
+
+class TwilioNotifier implements NotifierInterface
+{
+    protected $twilio
```
Comment (Crash--) [inline src/Notifications/TwilioNotifier.php]: ;

**Comment by jbrooksuk** [comment]
@Crash-- you can still put the data into tables, but you need to look doing that in a [seeder](http://laravel.com/docs/4.2/migrations#database-seeding), **not** the migration.

**Comment by Crash--** [comment]
@jbrooksuk Yes, but in the Cachet's documentation we can read " then you **may** want to seed the database with some example data." and from Laravel's "with **test** data using seed classes". But in my case, these datas are needed at the beginning. 

I can ofc change this behavior, but I'll add more lines of code :(

**Comment by Crash--** [comment]
@GrahamCampbell Can you close this PR? I'll open a new one with several fixes and improvements.

**Comment by jbrooksuk** [comment]
@Crash-- adding more lines of code isn't a problem if it's in the right place. If you're relying on default values to be inserted before the service can run, then you need to add error handling to only look for the service if the service is indeed setup.

**Comment by GrahamCampbell** [comment]
This needs rebasing and needs all language filled synced again.

**Comment by jbrooksuk** [comment]
Travis is failing because of a package not being able to download:

```
- Installing symfony/process (v2.5.8)
    Cloning 62c77d834c6cbf9cafa294a864aeba3a6c985af3
    Failed to download symfony/process from source: Failed to clone git@github.com:symfony/Process.git via git, https, ssh protocols, aborting.
```

**Comment by GrahamCampbell** [comment]
Hmm. I see nothing on github/travis status pages about this.

**Comment by jbrooksuk** [comment]
Also I saw "Could not authenticate github.com"

**Comment by jbrooksuk** [comment]
Do we have to add an API key or something (I've never had to).

**Comment by GrahamCampbell** [comment]
No, not at all.

**Comment by GrahamCampbell** [comment]
Travis has just f***ed up somewhere.

**Comment by jbrooksuk** [comment]
Bizarre. They need to run Cachet and let us know about this kind of shenanigans!

**Comment by GrahamCampbell** [comment]
Maybe we should tweet to them to let them know this is happening?

**Comment by GrahamCampbell** [comment]
> Bizarre. They need to run Cachet and let us know about this kind of shenanigans!

lol

**Comment by jbrooksuk** [comment]
Will do from the Cachet account.

**Comment by jbrooksuk** [comment]
Done :)

File: app/routes/dashboard.php [hunk 9]
```diff
@@ -66,8 +66,12 @@
     Route::group(['prefix' => 'notifications'], function () {
         Route::get('/', [
             'as'   => 'dashboard.notifications',
-            'uses' => 'DashboardController@showNotifications',
+            'uses' => 'NotificationController@showSlack',
         ]);
+        Route::get('slack', 'NotificationController@showSlack');
```
Comment (Crash--) [inline app/routes/dashboard.php]: I don't know how to redirect / to /slack. So actually, we have two routes to go at the same place...

Comment (jbrooksuk) [inline app/routes/dashboard.php]: How do you mean?

Comment (Crash--) [inline app/routes/dashboard.php]: L67 defines `/notifications` to be routed to `NotificationContoller@showSlack`
L71 defines `/notifications/slack` to be routed to the same function `NotificationController@showSlack` 

I'm not happy with that, I'd prefer to have only one route(`/notifications/slack`) to go to `NotificationController@showSlack` but if the user comes from /notifications I want to redirect it to /notifications/slack. A kind of alias I presume, but I don't know how to do that with Laravel's router.

Comment (jbrooksuk) [inline app/routes/dashboard.php]: Maybe a regular expression like this:

```php
Route::get('/{route}', [
    'as'   => 'dashboard.notifications',
    'uses' => 'NotificationController@showSlack',
])->where('route', '(slack)?');
```

**Comment by jbrooksuk** [comment]
Needs rebasing.

**Comment by jbrooksuk** [comment]
Rebased but not squashed yet. @GrahamCampbell please tell me that I didn't !@#$ it up this time?

**Comment by jbrooksuk** [comment]
@GrahamCampbell are you ok continuing work on this one? If not, what's left?

**Comment by joecohens** [comment]
@GrahamCampbell Maybe this will help https://github.com/dinkbit/notifyme

**Comment by jbrooksuk** [comment]
@joecohens how do you want to do this? Close this PR and implement fresh with notifyme, or modify this one?

**Comment by joecohens** [comment]
I think more than half of this branch is reusable. I'll take it form here :)

**Comment by joecohens** [comment]
@cachethq/owners Closing this for now, I'm starting fresh.

**Comment by jbrooksuk** [comment]
:+1:

**Pull request closed without merging**